src/v/storage/segment_appender.h

```diff
@@ -162,21 +205,76 @@ class segment_appender {
     // still heavy weight operations compared to regular flush()
     ss::future<> hard_flush();
 
+    enum write_state { QUEUED = 1, DISPATCHED, DONE };
+
     struct inflight_write {
```

Comment:
This struct is greatly expanded but is compensated by a corresponding reduction in the captured variables in the lambdas in dbhw. Overall I expect this new to be slightly faster as we avoid to repeated move/copy into a couple of lambdas (but that's not why I did it: it's necessary for the new semantics).